projects/plugins/boost/app/modules/optimizations/lcp/class-lcp-optimize-img-tag.php

```diff
@@ -0,0 +1,236 @@
+	 *
+	 * @since 4.0.0
+	 */
+	private function add_responsive_image_attributes( $element, $image_url ) {
+		if ( empty( $this->lcp_data['breakpoints'] ) ) {
+			return $element;
+		}
+
+		$srcset = $this->get_srcset( $image_url );
+		if ( ! empty( $srcset ) ) {
+			$element->set_attribute( 'srcset', $srcset );
+		}
+
+		$sizes = $this->get_sizes();
+		if ( ! empty( $sizes ) ) {
+			$element->set_attribute( 'sizes', $sizes );
+		}
+
+		return $element;
+	}
+
+	/**
+	 * Get the srcsets for an image.
+	 *
+	 * @param string $original_url The original image URL.
+	 * @return string The srcset for the image.
+	 *
+	 * @since $$next-version$$
+	 */
+	private function get_srcset( $original_url ) {
+		$widths = array();
+		foreach ( $this->lcp_data['breakpoints'] as $breakpoint ) {
+			$breakpoint_widths = array();
+			foreach ( $breakpoint['imageWidths'] as $width ) {
+				$breakpoint_widths[] = $width;
+
+				// If it's a Moto G Power, include a 1.75 DPR for accurate lighthouse representation of the optimized image.
+				if ( isset( $breakpoint['maxWidth'] ) && $breakpoint['maxWidth'] === 412 ) {
+					$breakpoint_widths[] = (int) $width * 1.75;
+				}
+
+				// Include 2x DPR.
+				$breakpoint_widths[] = $width * 2;
+
+				// If it's a mobile breakpoint, include 3x DPR.
+				if ( isset( $breakpoint['maxWidth'] ) && $breakpoint['maxWidth'] <= 480 ) {
+					$breakpoint_widths[] = $width * 3;
+				}
+			}
+			$widths[] = $breakpoint_widths;
+		}
+
+		$widths = array_unique( array_merge( ...$widths ) );
+
+		// Remove unnecessary widths to save some bytes in the HTML.
+		$widths = $this->reduce_widths( $widths );
+
+		$srcset = array();
+		foreach ( $widths as $width ) {
+			$srcset[] = Image_CDN_Core::cdn_url( $original_url, array( 'w' => $width ) ) . " {$width}w";
+		}
+
+		return implode( ', ', $srcset );
+	}
+
+	/**
+	 * Remove any width if we have another higher width that is within 20px.
+	 *
+	 * @param array $widths The widths to reduce.
+	 * @return array The reduced widths.
+	 */
+	private function reduce_widths( $widths ) {
+		rsort( $widths );
+		$reduced_widths = array();
+		$previous_width = 999999;
+		foreach ( $widths as $width ) {
+			if ( $previous_width - 20 < $width ) {
+				continue;
+			}
+
+			$previous_width   = $width;
+			$reduced_widths[] = $width;
+		}
+		return $reduced_widths;
+	}
+
+	/**
+	 * Get the sizes for an image.
+	 *
+	 * @return string The sizes for the image.
+	 *
+	 * @since $$next-version$$
+	 */
+	private function get_sizes() {
+		$sizes = array();
+		foreach ( $this->lcp_data['breakpoints'] as $breakpoint ) {
+			if ( empty( $breakpoint['widthValue'] ) ) {
+				continue;
+			}
+
+			// Make sure minWidth is valid if set.
+			if ( isset( $breakpoint['minWidth'] ) && ! is_numeric( $breakpoint['minWidth'] ) ) {
+				continue;
+			}
+
+			$min_width_query = isset( $breakpoint['minWidth'] ) ? '(min-width: ' . $breakpoint['minWidth'] . 'px) ' : '';
+
+			$sizes[] = $min_width_query . $breakpoint['widthValue'];
```

Comment:
Here we check if `minWidth` is validm but we don't do the same for `widthValue`. Should we?

Reply:
Added a regex based validation